### PR TITLE
Add a title attribute to a form

### PR DESCRIPTION
**Pull request by mhilton**

### CODE REVIEW

File: form/form.go [hunk 1]
```diff
@@ -31,6 +44,8 @@ type PromptingFiller struct {
 	MaxTries int
 }
 
+var _ Filler = &PromptingFiller{}
```
Comment (rogpeppe) [inline form/form.go]: (*PromptingFiller)(nil)

?

File: form/form_test.go [hunk 2]
```diff
@@ -147,9 +169,28 @@ func (s *formSuite) TestIOPrompter(c *gc.C) {
 	}
 }
 
+func (s *formSuite) TestIOPrompterWrtiteErrors(c *gc.C) {
```
Comment (rogpeppe) [inline form/form_test.go]: s/Wrtite/Write/

File: form/form.go [hunk 3]
```diff
@@ -42,24 +57,29 @@ type PromptingFiller struct {
...
 // giving up.
-func (f *PromptingFiller) Fill(fields environschema.Fields) (map[string]interface{}, error) {
-	fs := make(fieldSlice, 0, len(fields))
-	for k, v := range fields {
+func (f *PromptingFiller) Fill(form Form) (map[string]interface{}, error) {
+	prompter := f.Prompter
+	if prompter == nil {
+		prompter = DefaultPrompter
+	}
+	prompter.ShowTitle(form.Title)
```
Comment (rogpeppe) [inline form/form.go]: Perhaps this shouldn't call ShowTitle if form.Title is empty?

File: form/form.go [hunk 4]
```diff
@@ -126,10 +146,13 @@ func (f *PromptingFiller) prompt(name string, attr environschema.Attr) (interfac
 	return nil, errgo.New("too many invalid inputs")
 }
 
-// Prompt prompts for the value of the given named attribute.
-// It is always acceptable for an implementation to return a string,
-// which will be coerced to the appropriate type if possible.
 type Prompter interface {
+	// ShowTitle displays a form title.
```
Comment (rogpeppe) [inline form/form.go]: // It will only be called if the title of the form is non-empty.

?

File: form/form.go [hunk 5]
```diff
@@ -18,6 +18,19 @@ import (
 	"gopkg.in/juju/environschema.v1"
 )
 
+// Form describes a form based on a schema.
+type Form struct {
+	Title  string
```
Comment (rogpeppe) [inline form/form.go]: // Title holds the title of the form, giving contextual information
// for the fields.

?

File: form/form.go [hunk 6]
```diff
@@ -18,6 +18,19 @@ import (
...
+	Fields environschema.Fields
```
Comment (rogpeppe) [inline form/form.go:28]: // Fields holds the fields that make up the body of the form.

?

File: form/form.go (hunk 3 above)
Comment (rogpeppe) [inline form/form.go]: Also... we should probably check the error returned from ShowTitle.

File: form/form.go [hunk 7]
```diff
@@ -126,10 +146,13 @@ func (f *PromptingFiller) prompt(name string, attr environschema.Attr) (interfac
 	return nil, errgo.New("too many invalid inputs")
 }
 
-// Prompt prompts for the value of the given named attribute.
-// It is always acceptable for an implementation to return a string,
-// which will be coerced to the appropriate type if possible.
 type Prompter interface {
```
Comment (rogpeppe) [inline form/form.go:157]: // Prompter is the interface used by PromptingFiller.
// It is used to prompt the user for a sequence of form fields
// and obtain their values.
?

File: form/form.go [hunk 8]
```diff
@@ -222,3 +245,11 @@ func DefaultFromEnv(attr environschema.Attr) string {
 	}
 	return ""
 }
+
+// ShowTitle implements Prompter.ShowTitle.
```
Comment (rogpeppe) [inline form/form.go]: // ShowTitle implements Prompter.ShowTitle by printing
// the title to p.Out.

?

File: form/form.go [hunk 9]
```diff
@@ -147,7 +170,7 @@ type IOPrompter struct {
 	Out io.Writer
 }
 
-// Prompt implements Prompter. The prompt is generated from attr based on
+// Prompt implements Prompter.Prompt. The prompt is generated from attr based on
```
Comment (rogpeppe) [inline form/form.go]: // Prompt implements Prompter.Prompt by writing the field information
// to p.Out, then reading input from p.In. If p.In is a terminal and the
// attribute is secret, echo will be disabled.
?

File: form/form_test.go [hunk 10]
```diff
@@ -147,9 +169,28 @@ func (s *formSuite) TestIOPrompter(c *gc.C) {
...
+	return 0, errgo.New("test")
+}
+
 var promptingFillerTests = []struct {
 	about         string
-	schema        environschema.Fields
+	form          form.Form
 	maxTries      int
 	responses     []response
 	expectPrompts []prompt
```
Comment (rogpeppe) [inline form/form_test.go]: Shouldn't we have an expectShowTitle field here too, and a test or two that uses that?

File: form/form_test.go [hunk 11]
```diff
@@ -147,9 +169,28 @@ func (s *formSuite) TestIOPrompter(c *gc.C) {
...
+	prompter := form.IOPrompter{
+		In:  nil,
+		Out: errorWriter{},
+	}
+	err := prompter.ShowTitle("title")
+	c.Assert(err, gc.ErrorMatches, "cannot show title: test")
+	_, err = prompter.Prompt("attr", environschema.Attr{
+		Description: "A",
+	})
+	c.Assert(err, gc.ErrorMatches, "cannot write prompt: test")
```
Comment (rogpeppe) [inline form/form_test.go:184]: I'd like to see a test that PromptingFiller returns an error when ShowTitle does.

File: form/form_test.go [hunk 12]
```diff
@@ -612,8 +673,15 @@ type response struct {
...
-	prompts   []prompt
-	responses []response
+	title         string
+	titleResponse response
+	prompts       []prompt
+	responses     []response
+}
+
+func (p *testPrompter) ShowTitle(title string) error {
+	p.title = title
```
Comment (rogpeppe) [inline form/form_test.go]: perhaps p.showTitleCalled = true ?
then you'll be able to tell if it's been called appropriately or not.

**Comment by rogpeppe** [comment]
LGTM with a few minor issues for your consideration.

File: form/form.go (hunk 1 above)
Comment (mhilton) [inline form/form.go]: done

File: form/form_test.go (hunk 2 above)
Comment (mhilton) [inline form/form_test.go]: done

File: form/form.go (hunk 3 above)
Comment (mhilton) [inline form/form.go]: done

File: form/form.go (hunk 4 above)
Comment (mhilton) [inline form/form.go]: done

File: form/form.go (hunk 5 above)
Comment (mhilton) [inline form/form.go]: done

File: form/form.go (hunk 6 above)
Comment (mhilton) [inline form/form.go:28]: done

File: form/form.go (hunk 7 above)
Comment (mhilton) [inline form/form.go:157]: done

File: form/form.go (hunk 8 above)
Comment (mhilton) [inline form/form.go]: done

File: form/form.go (hunk 9 above)
Comment (mhilton) [inline form/form.go]: done

File: form/form_test.go (hunk 11 above)
Comment (mhilton) [inline form/form_test.go:184]: done

File: form/form_test.go (hunk 12 above)
Comment (mhilton) [inline form/form_test.go]: done, but in a slightly different way.

File: form/form_test.go (hunk 10 above)
Comment (mhilton) [inline form/form_test.go]: yes, done.

**Comment by mattyw** [comment]
LGTM

**Pull request merged**